### PR TITLE
Add transaction type filter to REST API 

### LINKED ISSUE #992: Support TransactionType Filter in RESTAPI w /api/v1/transactions?transactiontype
**Problem**
Currently The REST API for transactions only lets you filter by `account.id`, `timestamp`, `result (success/failure)` and `type (credit/debit)` - https://docs.hedera.com/guides/docs/mirror-node-api/cryptocurrency-api#transactions

This makes it challenging to query for say the 1st transaction that created an entity or a specific file operation.

It would be valuable to support filter by transaction type e.g. `CRYPTOCREATEACCOUNT` (11) or `CONSENSUSCREATETOPIC` (24)

**Solution**
Add an `/api/v1/transactions?transactiontype=<transactionTypeProtoId>` that filters transactions.
Consider whether this should allow a list of ids

**Alternatives**
- Refactor type to mean transactionType and make current `type (credit/debit)` be something like `entry (credit/debit)` or `credit (true/false)` 

**Additional Context**

### PR DESCRIPTION
**Pull request by Nana-EC**
**Detailed description**:
Currently The REST API for transactions only lets you filter by account.id, timestamp, result (success/failure) and type (credit/debit) - https://docs.hedera.com/guides/docs/mirror-node-api/cryptocurrency-api#transactions

This makes it challenging to query for say the 1st transaction that created an entity or a specific file operation.

- Add `getTransactionTypeQuery` to `utils.js` to create where clause for transaction filter
- Add `isValidTransactionType`  to `utils.js` to verify provided filter is in transactionType
- Add `transactionTypeQuery` to transaction innerQuery
- Add transactionTypes map to constants.js
- Add util tests for coverage
- Add spec files for coverage

**Which issue(s) this PR fixes**:
Fixes #992 

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: hedera-mirror-rest/utils.js [hunk 1]
```diff
@@ -724,6 +733,22 @@ const parsePublicKey = (publicKey) => {
...
+
+  const transactionType = parsedQueryParams[constants.filterKeys.TRANSACTION_TYPE];
+
+  if (isValidTransactionType(transactionType)) {
+    return `${constants.transactionColumns.TYPE}${opsMap.eq}${
+      constants.transactionTypes[transactionType.toUpperCase()]
+    }`;
+  }
+
+  return '';
```
Comment (steven-sheehy) [inline hedera-mirror-rest/utils.js:749]: I don't think it's a good user experience to silently fail and return all transactions if they mistype the transaction type. We should return an error if it's not one of the known types so they can adjust their criteria. Add a test for this scenario as well.

File: hedera-mirror-rest/__tests__/utils.test.js [hunk 2]
```diff
@@ -228,6 +228,33 @@ describe('Utils isValidNum tests', () => {
...
+    expect(utils.isValidTransactionType(-10)).toBe(false);
+  });
+  test('Verify invalid for 0', () => {
+    expect(utils.isValidTransactionType(0)).toBe(false);
+  });
+  test('Verify valid for valid CONSENSUSSUBMITMESSAGE transaction type', () => {
+    expect(utils.isValidTransactionType('CONSENSUSSUBMITMESSAGE')).toBe(true);
+  });
+  test('Verify valid for valid TOKENTRANSFERS transaction type', () => {
+    expect(utils.isValidTransactionType('TOKENTRANSFERS')).toBe(true);
```
Comment (steven-sheehy) [inline hedera-mirror-rest/__tests__/utils.test.js:254]: Add a lower case test.

File: hedera-mirror-rest/constants.js [hunk 3]
```diff
@@ -33,16 +34,22 @@ const filterKeys = {
   TIMESTAMP: 'timestamp',
   TOKENID: 'tokenId',
   TOKEN_ID: 'token.id',
-  TYPE: 'type',
+  CREDIT_TYPE: 'type',
+  TRANSACTION_TYPE: 'transactiontype',
```
Comment (steven-sheehy) [inline hedera-mirror-rest/constants.js:38]: Our naming is all over the map. Is this query param case sensitive? I think we need to use camel case `transactionType` to try to be consistent.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/constants.js (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-rest/constants.js:38]: I think right now it is case sensitive based on `filterValidityChecks()`
For a url query we should probably make all the filter keys case-insensitive when matching.

Comment (Nana-EC) [reply hedera-mirror-rest/constants.js:38]: There's a few ways to do this in a clean and easy to scale fashion 
1. Use a middle ware to ensure an expected case
2. Use an additional framework for clear request path and query param management. I think I have a ticket for this.

Either case is a larger and out of scope of this item. So I'll open a ticket to address that separately.

File: hedera-mirror-rest/utils.js (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-rest/utils.js:749]: Good point.
Will throw

File: hedera-mirror-rest/constants.js [hunk 4]
```diff
@@ -94,17 +101,58 @@ const defaultBucketNames = {
 
 const recordStreamPrefix = 'recordstreams/record';
 
+const transactionTypes = {
+  UNKNOWN: -1,
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-rest/constants.js:105]: Yeah, I can add another service for transaction types

[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-rest/constants.js:105]: Not a few line change and would add tests to make sure db query is good.
I'm proposing to move this to a v0.22 ticket to unblock v0.21

File: hedera-mirror-rest/__tests__/utils.test.js (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-rest/__tests__/utils.test.js:255]: Added

File: hedera-mirror-rest/constants.js (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-rest/constants.js]: As discussed went with hardcoding `transactionType` for now

File: hedera-mirror-rest/constants.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-rest/constants.js:105]: Moving this item to next release is fine

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Pull request merged**